internal/controllers/topology/cluster/desired_state.go

```diff
@@ -75,6 +76,13 @@ func (r *Reconciler) computeDesiredState(ctx context.Context, s *scope.Scope) (*
 	}
 	s.UpgradeTracker.MachineDeployments.MarkUpgrading(mdUpgradingNames...)
 
+	// Mark all the MachinePools that are currently upgrading.
+	mpUpgradingNames, err := s.Current.MachinePools.Upgrading(ctx, r.Client)
```

Comment:
Yup. We need the ClusterCacheTracker and get the client from there

Reply:
We have to do the following:
* Pass the tracker into the Cluster topology controllers:
  * Same as here: https://github.com/kubernetes-sigs/cluster-api/blob/8193dec816e6e3995d906620cb6fe557442d8e57/main.go#L495
  * here: https://github.com/kubernetes-sigs/cluster-api/blob/8193dec816e6e3995d906620cb6fe557442d8e57/main.go#L443
* Pass through the tracker in alias.go like this: https://github.com/kubernetes-sigs/cluster-api/blob/e222edfcabaf20d18fd52bda033344fca6073fef/controllers/alias.go#L103 (but for ClusterTopologyReconciler)
* And then here ~ something like this:
```go
	client, err := r.Tracker.GetClient(ctx, client.ObjectKeyFromObject(s.Current.Cluster)))
	if err != nil {
		return nil, err
	}
```

Additionally we have to make a change in the cluster topology controller
```go
	result, err :=  r.reconcile(ctx, s)
	if err != nil {
		// Requeue if the reconcile failed because the ClusterCacheTracker was locked for
		// the current cluster because of concurrent access.
		if errors.Is(err, remote.ErrClusterLocked) {
			log.V(5).Info("Requeuing because another worker has the lock on the ClusterCacheTracker")
			return ctrl.Result{Requeue: true}, nil
		}
	}
	return result, err
```
Here: https://github.com/kubernetes-sigs/cluster-api/blob/9d48366300f53c4ffd0da06925a6515678312573/internal/controllers/topology/cluster/cluster_controller.go#L204

This will also require some test changes, but let's make the changes to the production code first

Reply:
Just made these changes, thanks for such a detailed explanation! Just for my own understanding: Is the last part needed because to get the management cluster client, we need to access the ClusterCacheTracker but it does not have concurrent access?

Reply:
Roughly, yes. Basically multiple controllers (e.g Machine and Cluster controller) might try to get the client for a specific cluster at the same time. Usually this is very quick, but let's assume a workload cluster is not reachable. Basically getting a client will then mean trying to create a client and then timing out after 10 seconds.

We have a mechanism with TryLock in the CCT that at most one caller will try to create a client at the same time. All others will get this cluster locked error. If we wouldn't have that and let's assume we have a Cluster with 100 Machines. Basically this would mean that a lot of our workers of the Machine controller might just be stuck all trying to create a client for the same Cluster. So this is why we have the current implementation

We are now handling the error this way so that we just end up with a "clean" expontential backoff in those cases and not a lot of errors in our logs.

Reply:
Got it, thanks for the detailed explanation!

I'm having a bit of trouble figuring out where to get the ClusterCacheTracker from the Topology Plan [in this line](https://github.com/willie-yao/cluster-api/blob/96755a2311fbf8fdaa38ed710c53166f222aabd2/cmd/clusterctl/client/cluster/topology.go#L202). I don't think I have access to the main ClusterCacheTracker created in `main.go`. I noticed that the reconciler here is being setup for a dry run, so should we create another separate ClusterCacheTracker here?

Reply:
I did chat with @sbueringer and we came to the following conclusion:

* We cannot use the existing tracker implementation for clusterctl. Same applies for the test suite.
  * clusterctl: The management cluster and also maybe the cluster in the dry run may not even exist so there would be no api server to refer to at the `clusterctl alpha topology dryrun`.
  * test suite: It would require us to fake-create kubeconfig secrets to make it work, and these would need to point to reachable api server.


Because of that we propose to introduce and use a interface instead of `*remote.ClusterCacheTracker`. This way we can use different implementations for the above mentioned issues:
* For clusterctl: use a fakeclient based implementation for now. It is only used for `clusterctl alpha topology plan` and this way it for now would assume that it has to create the MachinePools.
* For the test suite: use a implementation which returns the envtest client. This way we could create node objects via the test suites which could be found.

Note: I would open up an issue to further build out the interface and use it across the codebase instead of directly refering the ClusterCacheTracker.

I created a PR on your fork to make it easier for you to review and adopt the proposed changes: https://github.com/willie-yao/cluster-api/pull/75

Reply:
Slightly more context. We have a solution in mind for clusterctl but we will do that in a follow-up PR. For now for this PR https://github.com/kubernetes-sigs/cluster-api/pull/9016#discussion_r1302948768 should be enough